runtime/tests/tests.rs

```diff
@@ -222,35 +263,318 @@ fn generic_asset_transfer_works_with_fee_exchange() {
+		.gas_price(1)
+		.build()
+		.execute_with(|| {
+			let contract_address = setup_contract(mock::contracts::CONTRACT_WITH_TRAP, alice());
+
+			// Call the newly instantiated contract. The contract is expected to dispatch a call
+			// and then trap.
+			let contract_call = Call::Contracts(pallet_contracts::Call::call(
+				contract_address, // newly created contract address
+				0,                // transfer value in
+				1_000_000,        // gas limit
+				vec![],
+			));
+			let contract_call_extrinsic = sign(CheckedExtrinsic {
+				signed: Some((bob(), signed_extra(2, 0, None, None))),
+				function: contract_call,
+			});
+
+			assert!(Executive::apply_extrinsic(contract_call_extrinsic).is_err());
+		});
+}
+
+#[test]
+fn contract_call_fails_with_insufficient_gas_without_fee_exchange() {
+	ExtBuilder::default()
+		.initial_balance(10_000 * TransactionBaseFee::get())
+		.gas_price(1)
+		.build()
+		.execute_with(|| {
+			initialize_block();
+			let xt = sign(CheckedExtrinsic {
+				signed: Some((alice(), signed_extra(0, 0, None, None))),
+				function: Call::Contracts(pallet_contracts::Call::call::<Runtime>(bob(), 10, 1000, vec![])),
+			});
+			assert_eq!(
+				Executive::apply_extrinsic(xt),
+				Ok(Err(DispatchError::Other("not enough gas to pay transfer fee")))
+			);
+
+			let xt_below_base_call_fee = sign(CheckedExtrinsic {
+				signed: Some((alice(), signed_extra(1, 0, None, None))),
+				function: Call::Contracts(pallet_contracts::Call::call::<Runtime>(bob(), 10, 1, vec![])),
+			});
+			assert_eq!(
+				Executive::apply_extrinsic(xt_below_base_call_fee),
+				Ok(Err(DispatchError::Other("not enough gas to pay base call fee")))
+			);
+		});
 }
 
 #[test]
-fn contract_fails_with_insufficient_gas() {
-	// Not enough gas to run contract
+fn contract_call_fails_with_insufficient_gas_with_fee_exchange() {
+	ExtBuilder::default()
+		.initial_balance(10_000 * TransactionBaseFee::get())
+		.gas_price(1)
+		.build()
+		.execute_with(|| {
+			let liquidity_core_amount = 100 * TransactionBaseFee::get();
+			let liquidity_asset_amount = 200;
+
+			let _ = CennzxSpot::add_liquidity(
+				Origin::signed(alice()),
+				CENNZ_ASSET_ID,
+				10, // min_liquidity
+				liquidity_asset_amount,
+				liquidity_core_amount,
+			);
+			let ex_key = (CENTRAPAY_ASSET_ID, CENNZ_ASSET_ID);
+			assert_eq!(CennzxSpot::get_liquidity(&ex_key, &alice()), liquidity_core_amount);
+
+			let fee_exchange = FeeExchange::V1(FeeExchangeV1 {
+				asset_id: CENNZ_ASSET_ID,
+				max_payment: 100_000_000,
+			});
+
+			initialize_block();
+			let xt_at_base_call_fee = sign(CheckedExtrinsic {
+				signed: Some((alice(), signed_extra(0, 0, None, Some(fee_exchange.clone())))),
+				function: Call::Contracts(pallet_contracts::Call::call::<Runtime>(bob(), 10, 1000, vec![])),
+			});
+			assert_eq!(
+				Executive::apply_extrinsic(xt_at_base_call_fee),
+				Ok(Err(DispatchError::Other("not enough gas to pay transfer fee")))
```

Comment:
Same issue as before, the user has enough CENNZ, but we set the `gas_limit` low - we need to give the user not enough CENNZ to pay for a high enough `gas_limit`, that way we are testing our `fill_gas` and `empty_gas` functions